platform/kube/inject/initializer.go

```diff
@@ -0,0 +1,389 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package inject
+
+// NOTE: This tool only exists because kubernetes does not support
+// dynamic/out-of-tree admission controller for transparent proxy
+// injection. This file should be removed as soon as a proper kubernetes
+// admission controller is written for istio.
+
+import (
+	"encoding/json"
+	"time"
+
+	"github.com/golang/glog"
+
+	appsv1beta1 "k8s.io/api/apps/v1beta1"
+	batchv1 "k8s.io/api/batch/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/api/extensions/v1beta1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/strategicpatch"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+
+	proxyconfig "istio.io/api/proxy/v1/config"
+	"istio.io/pilot/tools/version"
+)
+
+const (
+	initializerName = "sidecar.initializer.istio.io"
+	patchType       = types.StrategicMergePatchType
+)
+
+var ignoredNamespaces = []string{
+	"kube-system",
+	"kube-public",
+}
+
+// InitializerOptions stores the configurable options for an initializer
+type InitializerOptions struct {
+	// TODO - pull from ConfigMap?
+
+	// ResyncPeriod specifies how frequently to retrieve the full list
+	// of watched resources for initialization.
+	ResyncPeriod time.Duration
+
+	// Hub is the image registry used for the injected sidecar proxy
+	// and init images, e.g. docker.io/istio.
+	Hub string
+
+	// Tag is the docker image tag used for the injected sidecar proxy
+	// and init images.
+	Tag string
```

Comment:
Agreed. We went back and forth on these options with kube-inject and settled on hub/tag instead since users had to type it at the command line everyone. This is no longer an issue with initializer but I kept hub/tag for consistency with existing kube-inject. We could fix this as part of https://github.com/istio/pilot/issues/1062.